pkg/storageops/vsphere/vsphere.go

```diff
@@ -0,0 +1,397 @@
+package vsphere
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/kubernetes/kubernetes/pkg/cloudprovider/providers/vsphere/vclib/diskmanagers"
+	"github.com/libopenstorage/openstorage/pkg/storageops"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/govmomi/vim25/types"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib"
+)
+
+const (
+	diskDirectory  = "osd-provisioned-disks"
+	dummyDiskName  = "kube-dummyDisk.vmdk"
+	diskByIDPath   = "/dev/disk/by-id/"
+	diskSCSIPrefix = "wwn-0x"
+)
+
+type vsphereOps struct {
+	vm   *vclib.VirtualMachine
+	conn *vclib.VSphereConnection
+	cfg  *VSphereConfig
+}
+
+// NewClient creates a new vsphere storageops instance
+func NewClient(cfg *VSphereConfig) (storageops.Ops, error) {
+	vSphereConn := &vclib.VSphereConnection{
+		Username:          cfg.User,
+		Password:          cfg.Password,
+		Hostname:          cfg.VCenterIP,
+		Insecure:          cfg.InsecureFlag,
+		RoundTripperCount: cfg.RoundTripperCount,
+		Port:              cfg.VCenterPort,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	vmObj, err := GetVMObject(ctx, vSphereConn, cfg.VMUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	availableDatastores, err := vmObj.GetAllAccessibleDatastores(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create datastore URL map for faster lookup
+	datastores := make([]string, 0)
+	for _, ds := range availableDatastores {
+		if strings.HasPrefix(ds.Info.Name, cfg.Datastore) {
+			datastores = append(datastores, ds.Info.Name)
+		}
+	}
+
+	if len(datastores) == 0 {
+		return nil, fmt.Errorf("failed to find any available datastores to vm: %s for given prefix: %s", vmObj.Name(), cfg.Datastore)
+	}
+
+	logrus.Infof("Using following configuration for vsphere:")
+	logrus.Infof("  vCenter: %s:%s", cfg.VCenterIP, cfg.VCenterPort)
+	logrus.Infof("  Datacenter: %s", vmObj.Datacenter.Name())
+	logrus.Infof("  VMUUID: %s", cfg.VMUUID)
+	logrus.Infof("  Datastores: %v", datastores)
```

Comment:
Can the caller of Create API provide a datastore from this list or is it always going to be cfg.Datastore? If it is the latter case we should default to using cfg.Datastore instead of passing it in Create API.

Reply:
Caller always need to provide it in the create call. I've removed Datastore from cfg as it wasn't needed. The concept of datastores is only required by the consumers of this interface which is unit tests and portworx.